unity-renderer/Assets/Scripts/MainScripts/DCL/WorldRuntime/KernelCommunication/RPC/Services/SceneControllerService/SceneControllerServiceImpl.cs

```diff
@@ -0,0 +1,264 @@
+            CatalystSceneEntityMetadata parsedMetadata = Utils.SafeFromJson<CatalystSceneEntityMetadata>(request.Entity.Metadata);
+            Vector2Int[] parsedParcels = new Vector2Int[parsedMetadata.scene.parcels.Length];
+
+            for (int i = 0; i < parsedMetadata.scene.parcels.Length; i++)
+            {
+                parsedParcels[i] = Utils.StringToVector2Int(parsedMetadata.scene.parcels[i]);
+            }
+
+            await UniTask.SwitchToMainThread(ct);
+
+            if (request.IsGlobalScene)
+            {
+                CreateGlobalSceneMessage globalScene = new CreateGlobalSceneMessage()
+                {
+                    contents = parsedContent,
+                    id = request.Entity.Id,
+                    name = request.SceneName,
+                    baseUrl = request.BaseUrl,
+                    sceneNumber = sceneNumber,
+                    isPortableExperience = request.IsPortableExperience,
+                    icon = string.Empty // TODO: add icon url!
+                };
+
+                context.crdt.SceneController.CreateGlobalScene(globalScene);
+            }
+            else
+            {
+                LoadParcelScenesMessage.UnityParcelScene unityParcelScene = new LoadParcelScenesMessage.UnityParcelScene()
+                {
+                    sceneNumber = sceneNumber,
+                    id = request.Entity.Id,
+                    sdk7 = request.Sdk7,
+                    baseUrl = request.BaseUrl,
+                    baseUrlBundles = request.BaseUrlAssetBundles,
+                    basePosition = Utils.StringToVector2Int(parsedMetadata.scene.@base),
+                    parcels = parsedParcels,
+                    contents = parsedContent
+                };
+
+                context.crdt.SceneController.LoadUnityParcelScene(unityParcelScene);
+            }
+
+            LoadSceneResult result = new LoadSceneResult() { Success = true };
+            return result;
+        }
+
+        public async UniTask<UnloadSceneResult> UnloadScene(UnloadSceneMessage request, RPCContext context, CancellationToken ct)
+        {
+            // Debug.Log($"{GetHashCode()} SceneControllerServiceImpl.UnloadScene() - scene number: {sceneNumber}");
+
+            await UniTask.SwitchToMainThread(ct);
+            context.crdt.SceneController.UnloadParcelSceneExecute(sceneNumber);
+
+            return defaultUnloadSceneResult;
+        }
+
+        public async UniTask<CRDTSceneMessage> SendCrdt(CRDTSceneMessage request, RPCContext context, CancellationToken ct)
+        {
+            IParcelScene scene = null;
+            CRDTServiceContext crdtContext = context.crdt;
+
+            // This line is to avoid a race condition because a CRDT message could be sent before the scene was loaded
+            // more info: https://github.com/decentraland/sdk/issues/480#issuecomment-1331309908
+            await UniTask.WaitUntil(() => crdtContext.WorldState.TryGetScene(sceneNumber, out scene),
```

Comment:
In theory you may have concurrency issues with accessing a dictionary (that is behind `TryGetScene`) from a background thread.
I don't think it's needed to convert it to `ConcurrentDictionary` so I propose to move
```
await UniTask.SwitchToMainThread(ct);
```
before this line

Reply:
sounds good, tackled [here](https://github.com/decentraland/unity-renderer/pull/3700/commits/b415625c041167e4d80f2554ff61f531bbeda525)